### ISSUE TITLE
I have a question about the training process

### ISSUE THREAD

**Issue by 787629504** [karoly-hars/GAN_image_colorizing#1]
I'm sorry to disturb you. if I use your parameters to train the network, directly to train 200 epochs, do I need to adjust them in the process?

**Comment by karoly-hars** [comment]
When I was using regular batch norm, I trained with the following params:
```
python3 train.py --max_epoch=200 --smoothing=0.9 --l1_weight=0.99 --base_lr_gen=3e-4 \
--base_lr_disc=6e-5 --lr_decay_steps=6e4 --disc_norm=batch --apply_weight_init=1
```

and for spectral norm:
```
python3 train.py --max_epoch=100 --smoothing=1.0 --l1_weight=0.985 --base_lr_gen=2e-4 \
--base_lr_disc=2e-4 --lr_decay_steps=4e4 --disc_norm=spectral --apply_weight_init=0
```

(in both cases, for the rest of the params, I used the default values from ```train.py```)

I did not change any hyper params during training (other than the learning rate adjustments you can see line 105-108 in ```train.py```), so if you just run the code like this, you should get similar results as I did (at least on CIFAR10).

By the way, if you are looking for the trained models, you can download them from:
https://www.dropbox.com/s/r33ndl969q83gik/batchnorm_ep200_weigths_gen.pt
and
https://www.dropbox.com/s/tccxduyqp3dj5dg/spectralnorm_ep100_weights_gen.pt

**Comment by 787629504** [comment]
Thank you very much for your reply. 
I saw that the L1 loss parameter(MAE) was compared in that paper 《Image Colorization using Generative Adversarial Networks》, but when i run your code, it seems that i can't reach that goal(0.051). Have you compared with the parameter in the paper ?
(paper: MAE is computed by taking the mean of the absolute error of the generated and source images on a pixel level for each color channel)

**Comment by karoly-hars** [comment]
Some of the hyper params I used for training were different than to those described in the article. So if you are trying to achieve the same results as the authors, I would check out their repo:
https://github.com/ImagingLab/Colorizing-with-GANs

I remember that I initially tried to reach the same MAE as the article, but it was difficult the reproduce, or to even reach comparable results (in terms of MAE). After a while I let it go. A 5.1 MAE on CIFAR seemed like a totally unrealistic goal.

In general, I would say that MAE is not the best way to evaluate image colorizing. For example, if you increase the weight of the L1 loss in the training, the network will produce more grayish/sepia like images but the MAE loss might decrease.

**Comment by 787629504** [comment]
Thank you very much. And the value accuracy(24.1%,65.5%) ,can you reach that result? I wrote that code, but i can't reach that result.
(Accuracy is measured by the ratio between the number of pixels that have the same color information as the source and the total number of pixels.)

**Comment by karoly-hars** [comment]
I haven't tried. I ran the training sessions until I stopped seeing improvements in image quality. For the batch norm use case this was around 170-200 epochs and for spectral norm ~ 80-100 epochs.

**Comment by karoly-hars** [comment]
@787629504 
I did some digging and coding. I think there might be a misinterpretation about MAE on either my side or in the article. Maybe we are using a different subset or color space or something like that...

The mean absolute error is the same as the L1 loss, so in order to calculate the MAE, I am SUMMING up the ABSOLUTE difference of the real and generated pixel values of the test set (for every pixel on every image on every color channel) and I divide this sum by 10000x32x32x3 (num_images x height x width x num_channels).

Using this definition in the standard RGB color space (so the test image and the generated image both have pixel values from 0-255):
- the MAE of my batch norm model is around ~13.8
- the MAE of my spectral norm model is around ~12.7
- the MAE of the CIFAR10 model in the [article's repo](https://github.com/ImagingLab/Colorizing-with-GANs) is around ~14.5

*There might be some minor differences if we play around with the batch size during testing. For their code, I used a batch size of 128 to match the batch size in the CIFAR10 training.

Obviously, none of these are anywhere near the 5.1 MAE value stated in the article.

MAE on Wikipedia:
https://en.wikipedia.org/wiki/Mean_absolute_error

When I was working on this repo, I directed a question to the authors of the article since I figured maybe we are using MAE in different color spaces, but it seems like that is not the case:
https://github.com/ImagingLab/Colorizing-with-GANs/issues/13

So anyway, I am a little unsure what to do with all this. Maybe the MAE should be calculated differently in this case. :)

**Comment by 787629504** [comment]
Thank you very much for your reply. I understand a lot through your explanation. 

I saw that your code uses TORCH. nn. L1loss (reduction =' mean') to calculate L1 loss , i think it is the same as MAE loss value, but when I run the code, this value is usually around 0.07, There is a difference of 100 times from 13.8. I want to know why is this. Thank you!

**Comment by 787629504** [comment]
Sorry to disturb you. Are you still studying the direction of colorization?  Does the current evaluation standard use PSNR more? I read the two standards MAE and Accuracy in GAN's paper, which seem to have nothing to do with the quality of coloring. do you think that these two standards are important?

**Comment by karoly-hars** [comment]
**First your first question:**
Where did you see the 0.07 L1 loss? Is it printed during training? If so, during training the L1 loss is calculated on the training batches. In this phase, the images are in the LAB color space and they are normalized (all the pixel values are between -1 and 1). So the L1 loss here and the L1/MAE loss during the final evaluation on the test set mean something different.

**Regarding the second question:**
I am not an expert on image colorization. I just studied it for a few weeks to build this repo. I am not sure what loss function or eval method performs the best in the current state-of-the-art, but I think nor MAE nor accuracy are great ways to measure the capabilities of an image colorizing approach. The problem with these loss functions is that they are trying to quantify the difference between the original and the re-colorized image, but getting the exact same image as the original is not necessarily the goal here. I would say, the goal is the create a lifelike image that is "believable" to humans. This is basically why people started experimenting with GANs on this domain. Adversarial loss promotes plausibility over similarity.

Example: the sky is obviously usually blue and the grass is usually green, but the color of a car can be any color (more-or-less). When we are using MAE or accuracy, the network is punished for painting a car red if it was originally blue, but that in itself is not necessarily a mistake if the results are believable.
![Screenshot from 2020-02-16 11-11-05](https://user-images.githubusercontent.com/39346572/74602838-50992700-50ad-11ea-9c86-9e51035095e6.png)
_Left to right: original, grayscale(input), and the output of two different colorizing approaches._

On the image above, I would say that the networks did a good job with the colorizing, but the MAE is probably pretty high because they did not "guess" the original color of the car correctly.

**Comment by 787629504** [comment]
Thank you very much for your reply. I understand a lot through your reply.
